ext/ddtrace_profiling_native_extension/collectors_discrete_dynamic_sampler.c

```diff
@@ -0,0 +1,323 @@
+
+  // check if we should readjust our sampler after this event
+  maybe_readjust(sampler, now_ns);
+
+  if (should_sample) {
+    sampler->sample_start_time_ns = now_ns;
+  }
+
+  return should_sample;
+}
+
+bool discrete_dynamic_sampler_should_sample(discrete_dynamic_sampler *sampler) {
+  long now = monotonic_wall_time_now_ns(DO_NOT_RAISE_ON_FAILURE);
+  return _discrete_dynamic_sampler_should_sample(sampler, now);
+}
+
+static long _discrete_dynamic_sampler_after_sample(discrete_dynamic_sampler *sampler, long now_ns) {
+  long last_sampling_time_ns = sampler->sample_start_time_ns == 0 ? 0 : long_max_of(0, now_ns - sampler->sample_start_time_ns);
+  sampler->samples_since_last_readjustment++;
+  sampler->sampling_time_since_last_readjustment_ns += last_sampling_time_ns;
+  sampler->events_since_last_sample = 0;
+
+  // check if we should readjust our sampler after this sample
+  maybe_readjust(sampler, now_ns);
+
+  return last_sampling_time_ns;
+}
+
+long discrete_dynamic_sampler_after_sample(discrete_dynamic_sampler *sampler) {
+  long now = monotonic_wall_time_now_ns(DO_NOT_RAISE_ON_FAILURE);
+  return _discrete_dynamic_sampler_after_sample(sampler, now);
+}
+
+double discrete_dynamic_sampler_event_rate(discrete_dynamic_sampler *sampler) {
+  return sampler->events_per_ns * 1e9;
+}
+
+double discrete_dynamic_sampler_probability(discrete_dynamic_sampler *sampler) {
+  return sampler->sampling_probability * 100.;
+}
+
+long discrete_dynamic_sampler_sampling_time_ns(discrete_dynamic_sampler *sampler) {
+  return sampler->sampling_time_ns;
+}
+
+size_t discrete_dynamic_sampler_events_since_last_sample(discrete_dynamic_sampler *sampler) {
+  return sampler->events_since_last_sample;
+}
+
+double discrete_dynamic_sampler_target_overhead_adjustment(discrete_dynamic_sampler *sampler) {
+  return sampler->target_overhead_adjustment * 100.;
+}
+
+static void maybe_readjust(discrete_dynamic_sampler *sampler, long now) {
+  long window_time_ns = sampler->last_readjust_time_ns == 0 ? ADJUSTMENT_WINDOW_NS : now - sampler->last_readjust_time_ns;
+
+  if (window_time_ns < ADJUSTMENT_WINDOW_NS) {
+    // not enough time has passed to perform a readjustment
+    return;
+  }
+
+  // If we got this far, lets recalculate our sampling params based on new observations
+
+  // Update our running average of events/sec with latest observation
+  sampler->events_per_ns = EXP_MOVING_AVERAGE((double) sampler->events_since_last_readjustment / window_time_ns, sampler->events_per_ns);
+
+  // Update our running average of sampling time for a specific event
+  long sampling_window_time_ns = sampler->sampling_time_since_last_readjustment_ns;
+  if (sampler->samples_since_last_readjustment > 0) {
+    long avg_sampling_time_in_window_ns = sampler->samples_since_last_readjustment == 0 ? 0 : sampling_window_time_ns / sampler->samples_since_last_readjustment;
+    sampler->sampling_time_ns = EXP_MOVING_AVERAGE(avg_sampling_time_in_window_ns, sampler->sampling_time_ns);
```

Comment:
I don't think we should? Seems weird to try going for more precision than that of your actual data source (which is long-based). I also assume ops with longs are slightly more efficient than those with doubles (probably negligible in practice but...). Finally, it's more of an itch of knowing doubles can't actually store 64 bit values with enough precision. Granted for these two specific ones that's not an issue since they're essentially storing only diffs but I'd rather align all `_ns` variables on types.

Unless there's a good reason to do so that I'm missing?

Reply:
You make good points... :thinking: 

My reasoning for this suggestion is I was thinking about the rounding. For both `avg_sampling_time_in_window_ns` and `sampling_time_ns` we're doing a bunch of divisions and multiplications (by non-integers in the case of the moving average), so I was kinda wondering if they wouldn't result in more irregular behavior.

I'll readily admit it's more "I'm concerned that maybe" and less "I'm very sure about the issue". :rofl: 

Having said that... out of curiosity...

> Finally, it's more of an itch of knowing doubles can't actually store 64 bit values with enough precision.

Do note that [integers up to 2^53](https://stackoverflow.com/questions/1848700/biggest-integer-that-can-be-stored-in-a-double) can be stored without loss of precision. Since we're using ns, if my math doesn't fail me, 2^53 would be ~104 days which I wouldn't expect for any of these variables ;)